### PR TITLE
Add list element feature [Fixes #772]

### LINKED ISSUE #772: [Feature request] List group
I haven't found anything similar to Bootstrap 4's `list-group`.
image

Are you interested in this feature? If so, I can make a PR.

### PR DESCRIPTION
**Pull request by danieldaeschle**
This is a **new feature**.
_How to update the docs/changelog?_

### Proposed solution
It adds a list element which was missing of the most of users

### Testing Done
I checked in browser how it looks
image

### CODE REVIEW

**Comment by Log1x** [comment]
You forgot to remove `list.sass` from elements when you changed it to a component.

Otherwise, looks good to me.

**Comment by danieldaeschle** [comment]
Updated.

File: sass/components/list.sass [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+$list-radius: $radius !default
+
+$list-item-border: 1px solid $border !default
+$list-item-hover-color: $background !default
+
+.list
+  background-color: $list-background-color
+  border: $list-border
+  border-radius: $list-radius
+  .list-item
```
Comment (jgthms) [inline sass/components/list.sass]: I wouldn't nest this under `.list`.

File: sass/components/list.sass [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    color: black
```
Comment (jgthms) [inline sass/components/list.sass]: Use `color: $list-item-color` and set the default value to `$text`.

File: sass/components/list.sass [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    cursor: default
```
Comment (jgthms) [inline sass/components/list.sass]: Remove this.

File: sass/components/list.sass [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    display: flex
```
Comment (jgthms) [inline sass/components/list.sass]: Remove this.

File: sass/components/list.sass [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    justify-content: flex-start
```
Comment (jgthms) [inline sass/components/list.sass]: Remove this.

File: sass/components/list.sass [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    padding: 8px 16px
```
Comment (jgthms) [inline sass/components/list.sass]: Use `padding: 0.5em 1em`

File: sass/components/list.sass [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    &:not(:last-child)
+      border-bottom: $list-item-border
+  & .is-hoverable:hover
```
Comment (jgthms) [inline sass/components/list.sass]: I like this idea. But the syntax is `&.is-hoverable:hover` (without the space) and should be nested under `.list-item`.

Also, can you add a `&.is-active` state?

File: sass/components/list.sass (hunk 1 above)
Comment (danieldaeschle) [reply sass/components/list.sass]: where then?

File: sass/components/list.sass (hunk 2 above)
Comment (danieldaeschle) [reply sass/components/list.sass]: oh yes, my fault :O

File: sass/components/list.sass (hunk 6 above)
Comment (danieldaeschle) [reply sass/components/list.sass]: i didn't understand em and rem until today xD

File: sass/components/list.sass (hunk 4 above)
Comment (danieldaeschle) [reply sass/components/list.sass]: without `display: flex`
image

File: sass/components/list.sass (hunk 7 above)
Comment (danieldaeschle) [reply sass/components/list.sass]: which color should it have? i tried to give it primary but then the text color is still black

**Comment by KKSzymanowski** [comment]
With the current implementation card appears to be slightly wider then the list because of border instead of box shadow. Switching to box shadow fixes this and, in my opinion, gives a nicer feel to it:
```diff
.list
  background-color: $list-background-color
- border: $list-border
+ box-shadow: $card-shadow
  border-radius: $list-radius
  &.is-hoverable > .list-item:hover:not(.is-active)
    background-color: $list-item-hover-color
    cursor: pointer
```


image

**Comment by KKSzymanowski** [comment]
Also, one of my use-cases for this list would be either a key value list or a list of labels and icons:
```html
<div class="list">
    <div class="list-item">
        <div class="level">
            <div class="level-left">
                First
            </div>
            <div class="level-right">
                <i class="fa fa-check fa-fw"></i>
            </div>
        </div>
    </div>
</div>
```
image

But because of the `display: flex` on the `.list-item` the level element doesn't fill the entire width:

image

What would be the workaround for that?

**Comment by danieldaeschle** [comment]
eventually `display: block` is the solution? :)

**Comment by KKSzymanowski** [comment]
Yeah, that's how I did it for the screenshot, but I thought the flex display was there for a reason.

**Comment by danieldaeschle** [comment]
What does @jgthms think about the border?

**Comment by jgthms** [comment]
The shadow is a good shout. Let’s put that instead of the border.

**Comment by KKSzymanowski** [comment]
I haven't noticed this before but left padding is not consistent with the card.
image

Wouldn't it be better to change the list item padding to 0.75 and switch `em` to `rem`?

**Comment by danieldaeschle** [comment]
`padding: 0.5rem 0.75rem` then? or `padding: 0.5em 0.75rem` ? what is the different?

**Comment by KKSzymanowski** [comment]
In my opinion `padding: 0.5rem 0.75rem` (or even `padding: 0.375rem 0.75rem`).

> what is the different?

`em` refers to the subject's font size whereas `rem` refers to the root(`<html>`) font size.

File: sass/components/list.sass [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  box-shadow: $card-shadow
+  border-radius: $list-radius
+  &.is-hoverable > .list-item:hover:not(.is-active)
+    background-color: $list-item-hover-color
+    cursor: pointer
+
+.list-item
+  color: $list-item-color
+  display: block
+  padding: 0.375rem 0.75rem
```
Comment (jgthms) [inline sass/components/list.sass]: You need `padding: 0.5em 1em` here.

File: sass/components/list.sass [hunk 9]
```diff
@@ -0,0 +1,26 @@
+$list-background-color: $white !default
+$list-radius: $radius !default
+
+$list-item-border: 1px solid $border !default
+$list-item-color: $text !default
+$list-item-active-color: $primary !default
```
Comment (jgthms) [inline sass/components/list.sass]: `$list-item-active-background-color: $link !default`

File: sass/components/list.sass [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+$list-item-active-text-color: $white !default
```
Comment (jgthms) [inline sass/components/list.sass]: `$list-item-active-color: $link-invert !default`

File: sass/components/list.sass [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+$list-item-hover-color: $background !default
```
Comment (jgthms) [inline sass/components/list.sass]: `$list-item-hover-background-color: $background !default`

File: sass/components/list.sass (hunk 8 above)
Comment (KKSzymanowski) [reply sass/components/list.sass]: @jgthms Why is the `0.5em 1em` padding better than `0.375rem 0.75rem`?

Comment (jgthms) [reply sass/components/list.sass]: Uses `em`so it's proportional to the list's font size, and just looks better.

Comment (KKSzymanowski) [reply sass/components/list.sass]: I get the second part, but don't understand how do you decide whether to use `rem` or `em`.
For vertical margins(level, card, panel) you tend to use `rem` but sometimes(form label) you use `em`.
For horizontal margins you tend to use `em` but sometimes(tags) you use `rem`.
Similarly for paddings.
What's your rule of thumb when it comes to these units?

I feel like I'm being annoying but I just want to understand your train of though.

Comment (jgthms) [reply sass/components/list.sass]: Margin is between elements, hence a common denominator like rem. Padding is within an element and the visual balance is assured when everything is proportional to the font size hence em.

Button padding is em. Same for form controls.

**Comment by danieldaeschle** [comment]
Ready to merge?

File: sass/components/list.sass [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+$list-item-border: 1px solid $border !default
+$list-item-color: $text !default
+$list-item-active-background-color: $link !default
+$list-item-active-color: $link-invert !default
+$list-item-hover-background-color: $background !default
+
+.list
+  background-color: $list-background-color
+  box-shadow: $card-shadow
```
Comment (jgthms) [inline sass/components/list.sass]: Don’t reference a variable from another file. Create a new one.

Comment (danieldaeschle) [reply sass/components/list.sass]: xD

**Comment by danieldaeschle** [comment]
Feedback?

**Comment by danieldaeschle** [comment]
push

**Comment by HDVinnie** [comment]
Any update? Will this be merged?

**Comment by Tagueo** [comment]
Any updates ? I think this feature would be very usefull

**Comment by frederikhors** [comment]
@jgthms can you please merge? Is there another collaborator who can do it for you?

**Comment by frederikhors** [comment]
@danieldaeschle can you update docs?

**Comment by KKSzymanowski** [comment]
Do what you can and tag us, we'll help you out.

**Comment by danieldaeschle** [comment]
Thanks, but I'm waiting for confirmation from @jgthms

**Comment by danieldaeschle** [comment]
Wow, thanks :)

**Comment by jgthms** [comment]
No, thank you! 😉

**Comment by frederikhors** [comment]
@danieldaeschle any docs? How to use? With english we can help.

**Comment by KKSzymanowski** [comment]
@frederikhors Not much to it:
```html
<div class="list">
    <div class="list-item">First</div>
    <div class="list-item is-active">Second</div>
    <div class="list-item">Third</div>
</div>
```

If you want the items behaving like links:
```html
<div class="list is-hoverable">
    <div class="list-item">First</div>
    <div class="list-item is-active">Second</div>
    <div class="list-item">Third</div>
</div>
```

**Comment by nandordudas** [comment]
Link of list is missing at components, i try with url and it works.
https://bulma.io/documentation/components/list/

**Pull request merged**